### ISSUE TITLE
RSA PSS verify problem 

### ISSUE THREAD

**Issue by r06631041** [ARMmbed/mbedtls#4046]
I want to use NIST's test vector to test mbedtls lib. and I don’t know if the following program is wrong.
### Description
- Type: Bug
- Priority:  Major 

---------------------------------------------------------------

**mbed TLS build:**  
Configuration: 
/* System support */
#define MBEDTLS_PLATFORM_C

/* mbed TLS modules */
#define MBEDTLS_AES_C
#define MBEDTLS_SHA1_C
#define MBEDTLS_SHA256_C
#define MBEDTLS_SHA512_C

#define MBEDTLS_MD_C
#define MBEDTLS_OID_C
#define MBEDTLS_BIGNUM_C
#define MBEDTLS_RSA_C
#define MBEDTLS_PKCS1_V15
#define MBEDTLS_PKCS1_V21
#define MBEDTLS_AES_ROM_TABLES

-------------------------------------------------------------------------------------
## TEST vector  
```
SHAAlg =SHA256
[mod = 2048]
n = c6e0ed537a2d85cf1c4effad6419884d824ceabf5200e755691cb7328acd6a755fe85798502ccaec9e55d47afd0cf3258ebe920b50c5fd9d72897462bd0e459bbdf902b63d17195b2ef54908980be12aa7489f8af274b92c0cbc16aed2fa46f782d5517b666edfb2e5e5efeaff7e24965e26472e51980b0cfe457d297e6aa5dacb8e728dc6f58130f925a13275c3cace62f820db1e13cc5274c58ff4d7837671a1bf5f80d6ad8699c568df8d24dd0f152ded36ef4861f59b354bba96a076913a25facf4722737e6deed95b69a00fb2bced0feeedea4ff01a92605cfe26a6b39553d0c74e5650eb3779705e135c4b2fa518a8d4339c53efab4bb0058238def555
p = e2f7ceb13ea5385ad7659d7bbf0ad4a517c697b70c9d2af7a2193d62b14014412cc3e5fda97882341e0a370ce9f0f6c8149bb199d6f408b65d0524aecaf6e3fd7e3c35de940dc661ae17ddbdf57184e75bd2e9642401045ba48c7aee4abdc1caddca85fd064e80ab82ce58537848d9e9b8a477b4dfd04b9be496baec79cfa4c5
q = e0515147b6e596b7e5140e81365ad698dbeddd874642510f42d357123c20ffb0e1f377afefe97f20442e1c3f3c88919c39978b78835b9c7253f6ea632ab329866748c6dc195865ce123c8e153d03a3d731b7161205e2d83e6651152ee8181e389ad7a795dd3ce6ba44c753b4c7774fafbfa9c6606f89c08eec37632ba607b751
e = 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010001
d = 03ac73787e325992a96749d5ef8500e2ccf99e96214dbc22df2c6fde3538aaa8578e1b3cc871af5f940ed4a6df46438bdf240f896478fd2090fffa2af9c034a7cb684e5fc491f3940987c537d80128d6b37230ba4314c60d3580ad9aeb46ed6929dcf1629f6784667c547fec48c3112a1d9144f1802c82bb1476544e757e9853819185949352b92411adabd0f76efafe72c3b3fce88c5895b0bc4ac1ad36ec8d5be4adb89e72519850c6fc8c4076b658a2e554a37b5aa76aef7293a1ec256ccdc0c93c60aa528596a44ad72c76ed55726206d4bfd2f431745cc1c7dc399213051275fcfd2757552cef855be7bf23a5480688032bb4f322669a3e7d2fbff31c91
Msg = 81eaf473d40896dbf4deac0f35c63bd1e129147c76e7aa8d0ef921631f55a7436411079f1bcc7b98714ac2c13b5e7326e60d918db1f05ffb19da767a95bb141a84c4b73664ccebf844f3601f7c853f009b21becba11af3106f1de5827b14e9fac84b2cbf16d18c045622acb260024768e8acc4c0ae2c0bd5f60a98023828cdec
S = 40d59ebc6cb7b960cbda0db353f9b85d77e7c03f84447fb8e91b96a5a7377abc329d1f55c85e0dbedbc2886ce191d9e2cf3be05b33d6bbd2ba92b85eee2ff89cd6ee29cd531e42016e6aba1d620fe55e44480c033e8a59c0852dd1caffbc2ce82969e3a9f44ceff79f89993b9ebf3741b2ccab0b9516f2e128656a5b2ad5251e20c6ce0c26a14eef7ee86458942ddbe95ccc1f67b253e43e72117f49595dab5ba423496ece12825435661112666dbae71aaffd5a8f1d58db9dc02e0d70fe3ac36a87b8eeed4f20c00fd4303f9f767d03bca1a619bbe4b08e4e53b5cb69d2ba0235063e04ca392334d9979a41c42a66ca8b9721edcf76989ba89f3a170bb2e485
SaltVal = 11223344555432167890
Result = P
```
----------------------------------------------------------------
## code
```
   unsigned char Msg[] = "81eaf473d40896dbf4deac0f35c63bd1e129147c76e7aa8d0ef921631f55a7436411079f1bcc7b98714ac2c13b5e7326e60d918db1f05ffb19da767a95bb141a84c4b73664ccebf844f3601f7c853f009b21becba11af3106f1de5827b14e9fac84b2cbf16d18c045622acb260024768e8acc4c0ae2c0bd5f60a98023828cdec";
    unsigned char Sig[] = "40d59ebc6cb7b960cbda0db353f9b85d77e7c03f84447fb8e91b96a5a7377abc329d1f55c85e0dbedbc2886ce191d9e2cf3be05b33d6bbd2ba92b85eee2ff89cd6ee29cd531e42016e6aba1d620fe55e44480c033e8a59c0852dd1caffbc2ce82969e3a9f44ceff79f89993b9ebf3741b2ccab0b9516f2e128656a5b2ad5251e20c6ce0c26a14eef7ee86458942ddbe95ccc1f67b253e43e72117f49595dab5ba423496ece12825435661112666dbae71aaffd5a8f1d58db9dc02e0d70fe3ac36a87b8eeed4f20c00fd4303f9f767d03bca1a619bbe4b08e4e53b5cb69d2ba0235063e04ca392334d9979a41c42a66ca8b9721edcf76989ba89f3a170bb2e485";
    
    unsigned char n[] = "c6e0ed537a2d85cf1c4effad6419884d824ceabf5200e755691cb7328acd6a755fe85798502ccaec9e55d47afd0cf3258ebe920b50c5fd9d72897462bd0e459bbdf902b63d17195b2ef54908980be12aa7489f8af274b92c0cbc16aed2fa46f782d5517b666edfb2e5e5efeaff7e24965e26472e51980b0cfe457d297e6aa5dacb8e728dc6f58130f925a13275c3cace62f820db1e13cc5274c58ff4d7837671a1bf5f80d6ad8699c568df8d24dd0f152ded36ef4861f59b354bba96a076913a25facf4722737e6deed95b69a00fb2bced0feeedea4ff01a92605cfe26a6b39553d0c74e5650eb3779705e135c4b2fa518a8d4339c53efab4bb0058238def555";
    unsigned char e[] = "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010001";
    unsigned char d_Pkcsv21[] = "03ac73787e325992a96749d5ef8500e2ccf99e96214dbc22df2c6fde3538aaa8578e1b3cc871af5f940ed4a6df46438bdf240f896478fd2090fffa2af9c034a7cb684e5fc491f3940987c537d80128d6b37230ba4314c60d3580ad9aeb46ed6929dcf1629f6784667c547fec48c3112a1d9144f1802c82bb1476544e757e9853819185949352b92411adabd0f76efafe72c3b3fce88c5895b0bc4ac1ad36ec8d5be4adb89e72519850c6fc8c4076b658a2e554a37b5aa76aef7293a1ec256ccdc0c93c60aa528596a44ad72c76ed55726206d4bfd2f431745cc1c7dc399213051275fcfd2757552cef855be7bf23a5480688032bb4f322669a3e7d2fbff31c91";
    unsigned char p_Pkcsv21[] = "e2f7ceb13ea5385ad7659d7bbf0ad4a517c697b70c9d2af7a2193d62b14014412cc3e5fda97882341e0a370ce9f0f6c8149bb199d6f408b65d0524aecaf6e3fd7e3c35de940dc661ae17ddbdf57184e75bd2e9642401045ba48c7aee4abdc1caddca85fd064e80ab82ce58537848d9e9b8a477b4dfd04b9be496baec79cfa4c5";
    unsigned char q_Pkcsv21[] = "e0515147b6e596b7e5140e81365ad698dbeddd874642510f42d357123c20ffb0e1f377afefe97f20442e1c3f3c88919c39978b78835b9c7253f6ea632ab329866748c6dc195865ce123c8e153d03a3d731b7161205e2d83e6651152ee8181e389ad7a795dd3ce6ba44c753b4c7774fafbfa9c6606f89c08eec37632ba607b751";
    static unsigned char Msg_Hex[128] = { 0 };
    static unsigned char Sig_Hex[256] = { 0 };

    static unsigned char n_Hex[256] = { 0 };
    static unsigned char e_Hex[256] = { 0 };
    unsigned char d_Hex[256] = { 0 };
    unsigned char p_Hex[128] = { 0 };
    unsigned char q_Hex[128] = { 0 };
    int result=0;
    /*----------------------------------------*/
    TranslateACSIItoByte(Msg, Msg_Hex, 128);//convert string to Hex   example: "81eaf473d40" -> 0x81,0xea,0xf4,0x73....
    TranslateACSIItoByte(Sig, Sig_Hex, 256);
    TranslateACSIItoByte(n, n_Hex, 256);
    TranslateACSIItoByte(e, e_Hex, 256);
    TranslateACSIItoByte(d_Pkcsv21, d_Hex, 256);
    TranslateACSIItoByte(p_Pkcsv21, p_Hex, 128);
    TranslateACSIItoByte(q_Pkcsv21, q_Hex, 128);
    mbedtls_rsa_context rsa;
    mbedtls_rsa_init(&rsa, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA256);
    mbedtls_rsa_import_raw(&rsa, n_Hex, 256, p_Hex, 128, q_Hex, 128, d_Hex, 256, e_Hex, 256);
    unsigned char Msg_SHA256[256];
    SHA_256_plaintext(Msg_Hex, Msg_SHA256);// Plain text is SHA processing. 
    result=mbedtls_rsa_rsassa_pss_verify_ext(&rsa, NULL, NULL, MBEDTLS_RSA_PUBLIC, MBEDTLS_MD_SHA256,
        sizeof(Msg_SHA256), Msg_SHA256, MBEDTLS_MD_SHA256, MBEDTLS_RSA_SALT_LEN_ANY, Sig_Hex);
   

void TranslateACSIItoByte(char* u8Datain, char* u8Dataout, int u32length)
{
    int i;
    char u8temp1 = 0;
    char u8temp2 = 0;
    for (i = 0; i < u32length; i++)
    {
        //0~9
        if (u8Datain[i * 2 + 0] >= 0x30 && u8Datain[i * 2 + 0] <= 0x39)
        {
            u8temp1 = u8Datain[i * 2 + 0] - 0x30;
        }
        //A~F
        if (u8Datain[i * 2 + 0] >= 0x41 && u8Datain[i * 2 + 0] <= 0x46)
        {
            u8temp1 = u8Datain[i * 2 + 0] - 0x37;
        }
        //a~f
        if (u8Datain[i * 2 + 0] >= 0x61 && u8Datain[i * 2 + 0] <= 0x66)
        {
            u8temp1 = u8Datain[i * 2 + 0] - 0x57;
        }

        //0~9
        if (u8Datain[i * 2 + 1] >= 0x30 && u8Datain[i * 2 + 1] <= 0x39)
        {
            u8temp2 = u8Datain[i * 2 + 1] - 0x30;
        }
        //A~F
        if (u8Datain[i * 2 + 1] >= 0x41 && u8Datain[i * 2 + 1] <= 0x46)
        {
            u8temp2 = u8Datain[i * 2 + 1] - 0x37;
        }
        //a~f
        if (u8Datain[i * 2 + 1] >= 0x61 && u8Datain[i * 2 + 1] <= 0x66)
        {
            u8temp2 = u8Datain[i * 2 + 1] - 0x57;
        }

        u8Dataout[i] = ((u8temp1 & 0xf) << 4) | (u8temp2 & 0xf);
    }
}
void SHA_256_plaintext(unsigned char* MSG_real, unsigned char* res)
{
    printf("\nSHA256 Example:\n");
    unsigned char encrypt_Sha256[] = { 0 };
    unsigned char decrypt_Sha256[256];

    mbedtls_sha256_context sha256_context;
    mbedtls_sha256_init(&sha256_context);
    mbedtls_sha256_starts(&sha256_context, 0);
    mbedtls_sha256_update(&sha256_context, MSG_real, 128);
    mbedtls_sha256_finish(&sha256_context, decrypt_Sha256);
  
    for (int i = 0; i < 256; i++)
    {
        res[i] = decrypt_Sha256[i];
    }
   
    mbedtls_sha256_free(&sha256_context); //clear context
}
```
-----------------------------------------------------------------

## Question

I have found that PSS has salt values, but I did not import them when I did verify.
Although the result value from verify is equal to 0, I still want to confirm again.

Thanks.

**Comment by gilles-peskine-arm** [comment]
A PSS signature uses a random salt. To obtain exactly the same signature, you need to pass not only the same parameters (hash algorithm(s)), the same key and the same input message, but also the same salt. Most signature APIs, including the one in Mbed TLS, generate the random salt internally and don't have a parameter allowing the caller to use a specific salt. NIST test vectors have the salt spelled out so that the output is not randomized, but doing a known-answer test is not possible with implementations that aren't specifically designed for it (and that's why NIST allows compliance to only use sign-verify tests, or at least that was the case the last time I checked).

PSS verification decodes the salt from the signature, so it isn't necessary to pass it as an input. The verification _could_ allow the caller to pass an expected salt value, and check that the expected salt value matches the value that is decoded from the signature. But this check would have very little benefit (I'm not aware of any known attack that it would defend against), so most implementation, including Mbed TLS, don't implement such a check. Since the salt is normally unique, you'd have to transmit the salt value with the signature, which is usually not done, and would be subject to modification by a man-in-the-middle anyway.

Verifying the _length_ of the salt might increase the robustness somewhat (because the security proof for PSS requires a minimum salt size, so a signature made with a salt that's too short would be outside the scope of the proof), which is why it's moderately common for PSS verification implementations to allow the caller to specify the expected salt length. The salt length is normally a configuration parameter of the key, so it would be stored with the public key.

In summary, it's normal that the salt value is not useful to verify a PSS signature. This does not indicate either a defect in Mbed TLS or a defect in your code.

**Comment by gilles-peskine-arm** [comment]
This is not a bug report, but a support request, so I'm closing this issue.

If you think that Mbed TLS should offer a way to verify that an RSA PSS signature has a specific salt _value_ (and not just a specific salt length), please explain why, and we can reopen this issue. I'm not aware of any security reason or any protocol that requires it.
